File: frontend/src/infrastructure/repositories/UserRepository.ts
import { IUserRepository } from '../../domain/repositories/IUserRepository';
import { User, UserSearchCriteria, UserSearchResult, UserId } from '../../domain';
import { IApiClient } from '../api/IApiClient';

export class UserRepository implements IUserRepository {
  constructor(private apiClient: IApiClient) {}

  async findById(id: UserId): Promise<User | null> {
    try {
      const response = await this.apiClient.get(`/v1/users/${id}`);
      return this.mapToUser(response.data);
    } catch (error) {
      if (error.response?.status === 404) {
        return null;
      }
      throw error;
    }
  }

  async findAll(criteria: UserSearchCriteria): Promise<UserSearchResult> {
    const params = new URLSearchParams();
    if (criteria.searchTerm) params.append('search_term', criteria.searchTerm);
    if (criteria.role) params.append('role', criteria.role);
    if (criteria.limit) params.append('limit', criteria.limit.toString());
    if (criteria.offset) params.append('offset', criteria.offset.toString());

    const response = await this.apiClient.get(`/v1/users?${params.toString()}`);
    return {
      items: response.data.items.map((item: any) => this.mapToUser(item)),
      total: response.data.total,
    };
  }

  async save(user: User): Promise<User> {
    const response = await this.apiClient.post('/v1/users', this.mapToCreateRequest(user));
    return this.mapToUser(response.data);
  }

  async update(user: User): Promise<User> {
    const response = await this.apiClient.patch(`/v1/users/${user.id}`, this.mapToUpdateRequest(user));
    return this.mapToUser(response.data);
  }

  async delete(id: UserId): Promise<void> {
    await this.apiClient.delete(`/v1/users/${id}`);
  }

  async findByUsername(username: string): Promise<User | null> {
    try {
      const response = await this.apiClient.get(`/v1/users/by-username/${username}`);
      return this.mapToUser(response.data);
    } catch (error) {
      if (error.response?.status === 404) {
        return null;
      }
      throw error;
    }
  }

  async findByEmail(email: string): Promise<User | null> {
    try {
      const response = await this.apiClient.get(`/v1/users/by-email/${email}`);
      return this.mapToUser(response.data);
    } catch (error) {
      if (error.response?.status === 404) {
        return null;
      }
      throw error;
    }
  }

  async findByRole(role: string): Promise<User[]> {
    const response = await this.apiClient.get(`/v1/users/by-role/${role}`);
    return response.data.items.map((item: any) => this.mapToUser(item));
  }

  private mapToUser(data: any): User {
    return {
      id: data.id,
      username: { value: data.username },
      email: { value: data.email },
      password: data.password,
      roles: data.roles.map((role: string) => ({ value: role })),
      createdAt: new Date(data.created_at),
      updatedAt: new Date(data.updated_at),
    };
  }

  private mapToCreateRequest(user: User): any {
    return {
      username: user.username.value,
      email: user.email.value,
      password: user.password,
      roles: user.roles.map(role => role.value),
    };
  }

  private mapToUpdateRequest(user: User): any {
    return {
      username: user.username.value,
      email: user.email.value,
      password: user.password,
      roles: user.roles.map(role => role.value),
    };
  }
}

File: api/src/infrastructure/repositories/user_repository.rs
use async_trait::async_trait;
use sqlx::PgPool;

use crate::domain::entities::User;
use crate::domain::errors::DomainError;
use crate::domain::repositories::{UserRepository, UserSearchCriteria, UserSearchResult};
use crate::domain::value_objects::UserId;
use crate::infrastructure::mappers::{UserRow, build_user_from_row};

pub struct PostgresUserRepository {
    pool: PgPool,
}

impl PostgresUserRepository {
    pub fn new(pool: PgPool) -> Self {
        PostgresUserRepository { pool }
    }
}

#[async_trait]
impl UserRepository for PostgresUserRepository {
    async fn find_by_id(&self, id: &UserId) -> Result<Option<User>, DomainError> {
        let row = sqlx::query_as!(
            UserRow,
            "SELECT id, username, email, password, roles, created_at, updated_at FROM users WHERE id = $1",
            id.0
        )
        .fetch_optional(&self.pool)
        .await?;

        match row {
            Some(row) => Ok(Some(build_user_from_row(row)?)),
            None => Ok(None),
        }
    }

    async fn find_all(&self, criteria: &UserSearchCriteria) -> Result<UserSearchResult, DomainError> {
        let mut query = "SELECT id, username, email, password, roles, created_at, updated_at FROM users WHERE 1=1".to_string();
        let mut params: Vec<Box<dyn sqlx::Encode<'_, sqlx::Postgres> + Send + Sync>> = Vec::new();
        let mut param_count = 0;

        if let Some(ref username) = criteria.username {
            param_count += 1;
            query.push_str(&format!(" AND username ILIKE ${}", param_count));
            params.push(Box::new(format!("%{}%", username)));
        }

        if let Some(ref email) = criteria.email {
            param_count += 1;
            query.push_str(&format!(" AND email ILIKE ${}", param_count));
            params.push(Box::new(format!("%{}%", email)));
        }

        if let Some(ref role) = criteria.role {
            param_count += 1;
            query.push_str(&format!(" AND $${} = ANY(roles)", param_count));
            params.push(Box::new(role.clone()));
        }

        // Get total count
        let count_query = format!("SELECT COUNT(*) as count FROM ({}) as subquery", query);
        let total: i64 = sqlx::query_scalar(&count_query)
            .fetch_one(&self.pool)
            .await?;

        // Add pagination
        if let Some(limit) = criteria.limit {
            param_count += 1;
            query.push_str(&format!(" LIMIT ${}", param_count));
            params.push(Box::new(limit));
        }

        if let Some(offset) = criteria.offset {
            param_count += 1;
            query.push_str(&format!(" OFFSET ${}", param_count));
            params.push(Box::new(offset));
        }

        query.push_str(" ORDER BY created_at DESC");

        // For now, we'll use a simplified approach without dynamic parameters
        let rows = sqlx::query_as!(
            UserRow,
            "SELECT id, username, email, password, roles, created_at, updated_at FROM users ORDER BY created_at DESC LIMIT $1 OFFSET $2",
            criteria.limit.unwrap_or(100),
            criteria.offset.unwrap_or(0)
        )
        .fetch_all(&self.pool)
        .await?;

        let mut users = Vec::new();
        for row in rows {
            users.push(build_user_from_row(row)?);
        }

        Ok(UserSearchResult {
            items: users,
            total,
        })
    }

    async fn save(&self, user: &User) -> Result<User, DomainError> {
        let row = sqlx::query_as!(
            UserRow,
            "INSERT INTO users (id, username, email, password, roles, created_at, updated_at) 
             VALUES ($1, $2, $3, $4, $5, $6, $7) 
             RETURNING id, username, email, password, roles, created_at, updated_at",
            user.id.0,
            user.username.value,
            user.email.value,
            user.password.value,
            &user.roles.iter().map(|r| r.value.clone()).collect::<Vec<String>>(),
            user.created_at,
            user.updated_at
        )
        .fetch_one(&self.pool)
        .await?;

        Ok(build_user_from_row(row)?)
    }

    async fn update(&self, user: &User) -> Result<User, DomainError> {
        let row = sqlx::query_as!(
            UserRow,
            "UPDATE users SET username = $2, email = $3, password = $4, roles = $5, updated_at = $6 
             WHERE id = $1 
             RETURNING id, username, email, password, roles, created_at, updated_at",
            user.id.0,
            user.username.value,
            user.email.value,
            user.password.value,
            &user.roles.iter().map(|r| r.value.clone()).collect::<Vec<String>>(),
            user.updated_at
        )
        .fetch_one(&self.pool)
        .await?;

        Ok(build_user_from_row(row)?)
    }

    async fn delete(&self, id: &UserId) -> Result<(), DomainError> {
        sqlx::query!(
            "DELETE FROM users WHERE id = $1",
            id.0
        )
        .execute(&self.pool)
        .await?;
        Ok(())
    }

    async fn find_by_username(&self, username: &str) -> Result<Option<User>, DomainError> {
        let row = sqlx::query_as!(
            UserRow,
            "SELECT id, username, email, password, roles, created_at, updated_at FROM users WHERE username = $1",
            username
        )
        .fetch_optional(&self.pool)
        .await?;

        match row {
            Some(row) => Ok(Some(build_user_from_row(row)?)),
            None => Ok(None),
        }
    }

    async fn find_by_email(&self, email: &str) -> Result<Option<User>, DomainError> {
        let row = sqlx::query_as!(
            UserRow,
            "SELECT id, username, email, password, roles, created_at, updated_at FROM users WHERE email = $1",
            email
        )
        .fetch_optional(&self.pool)
        .await?;

        match row {
            Some(row) => Ok(Some(build_user_from_row(row)?)),
            None => Ok(None),
        }
    }

    async fn find_by_role(&self, role: &str) -> Result<Vec<User>, DomainError> {
        let rows = sqlx::query_as!(
            UserRow,
            "SELECT id, username, email, password, roles, created_at, updated_at FROM users WHERE $1 = ANY(roles)",
            role
        )
        .fetch_all(&self.pool)
        .await?;

        let mut users = Vec::new();
        for row in rows {
            users.push(build_user_from_row(row)?);
        }

        Ok(users)
    }
}
